### ISSUE TITLE
Unable to detect ghosttasks

### ISSUE THREAD

**Issue by j-wsy** [last-byte/PersistenceSniper#25]
Hi, love the tool.

I have a vm with a working ghosttask loaded (restarted and all that, and confirmed the ghosttask is functioning). I can see it in my regedit \tasks\, and I can see it doesn't have a SD, as expected. I can see it doing the action I made it do.

Running the latest PersistenceSniper v1.16.0, I can't seem to detect it. I see a bunch of other false-positives, so the tool itself is working, but no output relating to ghosttask.

**Comment by j-wsy** [comment]
![image](https://github.com/last-byte/PersistenceSniper/assets/44336384/958894a8-f419-4de4-a8c2-0f4d65ddd115)

**Comment by j-wsy** [comment]
Here's another example using a ghosttask named "April1", on a separate clean vm. Could not detect using Find-AllPersistence.

Am I doing something wrong/do I need to do anything else?

![postrebootScreenshot 2024-04-08 235441](https://github.com/last-byte/PersistenceSniper/assets/44336384/5e4fef0f-b4da-4371-be1e-2c140d9b0c4d)

**Comment by last-byte** [comment]
Hey there, solid copy. I’ll investigate it right away and fix it in the next minor release.

**Comment by j-wsy** [comment]
thanks, looking forward to it.

**Comment by daniele777** [comment]
Malware is fake amsi provider?

**Comment by strassi** [comment]
@j-wsy 

How did you create the ghost task. Some tools are creating ghosttask by deleting the SD (security descriptor) for the task, which hides it for every user.
![image](https://github.com/last-byte/PersistenceSniper/assets/14335437/42296125-625b-47f6-a0dd-19b291bc46a8)

But some tooling might preserve the SD value, which causes the persistencesniper script to fail finding the task. Persistence Sniper seems to check if the SD is missing. https://github.com/last-byte/PersistenceSniper/blob/6687a26250a3577a0f4c709cbf7c29191a5ad1e8/PersistenceSniper/PersistenceSniper.psm1#L1907 

Can you confirm of having a SD for your ghosttask? You have to search in the "Tree" Key for the specific ghost task.
<img width="400" alt="Fig4_Deletion-of-the-security-descriptor-SD-value" src="https://github.com/last-byte/PersistenceSniper/assets/14335437/7c7eaade-868e-41e8-bb4d-2d940db35551">

**Comment by j-wsy** [comment]
Hi @strassi 
I understand what you mean, though I have some questions.
![image](https://github.com/last-byte/PersistenceSniper/assets/44336384/551d1853-9a66-4736-a450-64a47282207c)

In green highlight, it shows the GhostTask "demo" with a \TaskCache\Tree\ SD. This was created via the method on https://github.com/netero1010/GhostTask - under "Example 1". Understandably after reading the PersistenceSniper source code, this would fail the boolean:  if ($null -eq $taskSD).

In netero1010's project, it describes the code as "Creates scheduled tasks **with a restrictive security descriptor**, making them invisible to all users."

It seems then that a GhostTask isn't about a _missing_ SD, but something within the SD that permits/denies visibility. So perhaps hunting for a missing SD is not right, but I don't have a better idea at the moment.

As a side note, I want to point out that if you follow the GhostTask's Id (in yellow highlight) and view it under \TaskCache\Tasks\, it has no SecurityDescriptor. I'm not sure if this is different to "SD" though, and also other IDs are missing SecurityDescriptor sometimes too. Regardless, once again a lack of an SD doesn't seem to imply a malicious GhostTask...

**Comment by strassi** [comment]
A possible workaround would be the detection of the used SD of the tool. The very restrictive permissions of GhostTask look like this:
https://github.com/netero1010/GhostTask/blob/7db8cd4120328a80e27e220785a8546bb1a4ed28/GhostTask.c#L769

**Comment by last-byte** [comment]
This should be fixed in 1.16.1 which I just released and seems to catch the use case :) @j-wsy @strassi

**Comment by j-wsy** [comment]
Awesome, thanks @last-byte 

It does work on the ghosttask I set up above using https://github.com/netero1010/GhostTask. 

I have a question though:

It looks like you added maliciousSD and it's comparing it to the SD (if there is one) of the suspected ghosttask. Where did you get this maliciousSD base64 encode from, and is this a hardcode value specific to the ghosttask SD in https://github.com/netero1010/GhostTask/blob/main/GhostTask.c#L769? 

It looks like that specific line `ConvertStringSecurityDescriptorToSecurityDescriptorA("O:BAG:SYD:", 1, &pSd, &sdLength);` is what generates the ghosttask SD data (https://learn.microsoft.com/en-us/windows/win32/secauthz/security-descriptor-string-format), and the D: part I think makes it restrictive? 

If we made a slight change to the unicode/hex of the SD, I guess it would bypass detection?
